File: frontend/src/types/detection.ts
/**
 * Types for cell detection API responses and application state.
 */

export interface BoundingBox {
  x_min: number;
  y_min: number;
  x_max: number;
  y_max: number;
}

/**
 * WBC Subtypes for differential classification
 */
export type WBCSubtype = 'Neutrophil' | 'Lymphocyte' | 'Monocyte' | 'Eosinophil' | 'Basophil';

/**
 * All cell types including WBC subtypes
 */
export type CellType = 'RBC' | 'Platelets' | WBCSubtype;

/**
 * Legacy cell type for backward compatibility
 */
export type LegacyCellType = 'RBC' | 'WBC' | 'Platelets';

export interface CellDetection {
  cell_type: CellType;
  confidence: number;
  bounding_box: BoundingBox;
}

/**
 * Cell counts from the API response with WBC differential
 */
export interface CellCounts {
  rbc: number;
  platelets: number;
  neutrophil: number;
  lymphocyte: number;
  monocyte: number;
  eosinophil: number;
  basophil: number;
}

/**
 * Helper to get total WBC count from all subtypes
 */
export const getTotalWBC = (counts: CellCounts): number => {
  return counts.neutrophil + counts.lymphocyte + counts.monocyte + counts.eosinophil + counts.basophil;
};

/**
 * Check if a cell type is a WBC subtype
 */
export const isWBCSubtype = (cellType: CellType): cellType is WBCSubtype => {
  return ['Neutrophil', 'Lymphocyte', 'Monocyte', 'Eosinophil', 'Basophil'].includes(cellType);
};

export interface DetectionResponse {
  success: boolean;
  message: string;
  counts: CellCounts;
  detections: CellDetection[];
  image_width: number;
  image_height: number;
}

export interface HealthResponse {
  status: string;
  version: string;
  model_loaded: boolean;
}

/**
 * Visibility state for all cell types including WBC subtypes
 */
export interface VisibleCellTypes {
  RBC: boolean;
  Platelets: boolean;
  Neutrophil: boolean;
  Lymphocyte: boolean;
  Monocyte: boolean;
  Eosinophil: boolean;
  Basophil: boolean;
}

/**
 * Color palette for cell types (color-blind friendly)
 * Based on UX Design Spec with WBC subtype variations
 */
export const CELL_TYPE_COLORS: Record<CellType, string> = {
  RBC: '#E07A5F',        // Coral Red
  Platelets: '#F2CC8F',  // Amber Yellow
  Neutrophil: '#3D5A80', // Ocean Blue (primary WBC color - most common)
  Lymphocyte: '#7C3AED', // Purple/Violet
  Monocyte: '#0D9488',   // Teal/Cyan
  Eosinophil: '#EA580C', // Orange
  Basophil: '#DB2777',   // Magenta/Pink
};

/**
 * Display labels for cell types
 */
export const CELL_TYPE_LABELS: Record<CellType, string> = {
  RBC: 'Red Blood Cells',
  Platelets: 'Platelets',
  Neutrophil: 'Neutrophils',
  Lymphocyte: 'Lymphocytes',
  Monocyte: 'Monocytes',
  Eosinophil: 'Eosinophils',
  Basophil: 'Basophils',
};

/**
 * Short labels for cell types
 */
export const CELL_TYPE_SHORT_LABELS: Record<CellType, string> = {
  RBC: 'RBC',
  Platelets: 'PLT',
  Neutrophil: 'NEU',
  Lymphocyte: 'LYM',
  Monocyte: 'MON',
  Eosinophil: 'EOS',
  Basophil: 'BAS',
};

File: backend/app/schemas/detection.py
"""
Pydantic schemas for cell detection API.

Defines request and response models for the detection endpoints.
"""

from pydantic import BaseModel, Field


class BoundingBox(BaseModel):
    """Bounding box coordinates for a detected cell."""

    x_min: float = Field(..., description="Left edge x-coordinate")
    y_min: float = Field(..., description="Top edge y-coordinate")
    x_max: float = Field(..., description="Right edge x-coordinate")
    y_max: float = Field(..., description="Bottom edge y-coordinate")


class CellDetection(BaseModel):
    """Single cell detection result."""

    cell_type: str = Field(
        ...,
        description="Type of cell detected (RBC, Platelets, Neutrophil, Lymphocyte, Monocyte, Eosinophil, Basophil)"
    )
    confidence: float = Field(..., ge=0.0, le=1.0, description="Detection confidence score")
    bounding_box: BoundingBox = Field(..., description="Bounding box coordinates")


class CellCounts(BaseModel):
    """Aggregated cell counts by type."""

    rbc: int = Field(default=0, description="Red blood cell count")
    platelets: int = Field(default=0, description="Platelet count")
    # WBC subtypes (differential classification)
    neutrophil: int = Field(default=0, description="Neutrophil count")
    lymphocyte: int = Field(default=0, description="Lymphocyte count")
    monocyte: int = Field(default=0, description="Monocyte count")
    eosinophil: int = Field(default=0, description="Eosinophil count")
    basophil: int = Field(default=0, description="Basophil count")

    @property
    def total_wbc(self) -> int:
        """Calculate total WBC count from all subtypes."""
        return self.neutrophil + self.lymphocyte + self.monocyte + self.eosinophil + self.basophil


class DetectionResponse(BaseModel):
    """Response model for cell detection endpoint."""

    success: bool = Field(..., description="Whether detection was successful")
    message: str = Field(..., description="Status message")
    counts: CellCounts = Field(..., description="Aggregated cell counts")
    detections: list[CellDetection] = Field(
        default_factory=list, description="List of individual detections"
    )
    image_width: int = Field(..., description="Width of processed image")
    image_height: int = Field(..., description="Height of processed image")


class HealthResponse(BaseModel):
    """Response model for health check endpoint."""

    status: str = Field(..., description="Service health status")
    version: str = Field(..., description="API version")
    model_loaded: bool = Field(..., description="Whether the ML model is loaded")
